src/components/tests/SourcesTree.spec.js

```diff
@@ -85,4 +86,622 @@ describe("SourcesTree", () => {
+      contents: new Map(),
+      name: "root",
+      path: "root/"
+    };
+
+    beforeEach(() => {
+      const rendered = render({
+        clearProjectDirectoryRoot: jest.fn(),
+        setProjectDirectoryRoot: jest.fn(),
+        projectRoot: "root/"
+      });
+      component = rendered.component;
+      props = rendered.props;
+    });
+
+    afterEach(() => {
+      showMenu.mockClear();
+      event.preventDefault.mockClear();
+      event.stopPropagation.mockClear();
+    });
+
+    it("shows context menu on directory to set as root", async () => {
+      const menuOptions = [
+        {
+          accesskey: "r",
+          click: expect.any(Function),
+          disabled: false,
+          id: "node-set-directory-root",
+          label: "Set directory root"
+        }
+      ];
+      await component.instance().onContextMenu(event, directory);
+      expect(showMenu).toHaveBeenCalledWith(event, menuOptions);
+
+      expect(event.preventDefault).toHaveBeenCalled();
+      expect(event.stopPropagation).toHaveBeenCalled();
+
+      showMenu.mock.calls[0][1][0].click();
+      expect(props.setProjectDirectoryRoot).toHaveBeenCalled();
+      expect(props.clearProjectDirectoryRoot).not.toHaveBeenCalled();
+      expect(copyToTheClipboard).not.toHaveBeenCalled();
+    });
+
+    it("shows context menu on file to copy source uri", async () => {
+      const menuOptions = [
+        {
+          accesskey: "u",
+          click: expect.any(Function),
+          disabled: false,
+          id: "node-menu-copy-source",
+          label: "Copy source URI"
+        }
+      ];
+      await component.instance().onContextMenu(event, file);
+      expect(showMenu).toHaveBeenCalledWith(event, menuOptions);
+
+      expect(event.preventDefault).toHaveBeenCalled();
+      expect(event.stopPropagation).toHaveBeenCalled();
+
+      showMenu.mock.calls[0][1][0].click();
+      expect(props.setProjectDirectoryRoot).not.toHaveBeenCalled();
+      expect(props.clearProjectDirectoryRoot).not.toHaveBeenCalled();
+      expect(copyToTheClipboard).toHaveBeenCalled();
+    });
+
+    it("shows context menu on root to remove directory root", async () => {
+      const menuOptions = [
+        {
+          click: expect.any(Function),
+          disabled: false,
+          id: "node-remove-directory-root",
+          label: "Remove directory root"
+        }
+      ];
+      await component.instance().onContextMenu(event, rootDirectory);
+      expect(showMenu).toHaveBeenCalledWith(event, menuOptions);
+
+      expect(event.preventDefault).toHaveBeenCalled();
+      expect(event.stopPropagation).toHaveBeenCalled();
+
+      showMenu.mock.calls[0][1][0].click();
+      expect(props.setProjectDirectoryRoot).not.toHaveBeenCalled();
+      expect(props.clearProjectDirectoryRoot).toHaveBeenCalled();
+      expect(copyToTheClipboard).not.toHaveBeenCalled();
+    });
+  });
+
+  describe("renderItem", () => {
+    const { component, props } = render();
```

Comment:
this is a bit dangerous. in future modifications someone might miss that this isn't in a before each.

I would however always colocate this kind of setup in the test itself. it makes it more difficult to make a mistake, and mistakes will be made.